extmod/vfs_posix.c

```diff
@@ -107,7 +112,28 @@ STATIC mp_obj_t vfs_posix_make_new(const mp_obj_type_t *type, size_t n_args, siz
     mp_obj_vfs_posix_t *vfs = mp_obj_malloc(mp_obj_vfs_posix_t, type);
     vstr_init(&vfs->root, 0);
     if (n_args == 1) {
-        vstr_add_str(&vfs->root, mp_obj_str_get_str(args[0]));
+        const char *root = mp_obj_str_get_str(args[0]);
+        // if the root is relative, make it absolute, otherwise we'll get confused by chdir
+        #ifdef _WIN32
+        char buf[MICROPY_ALLOC_PATH_MAX + 1];
+        DWORD result = GetFullPathNameA(root, sizeof(buf), buf, NULL);
+        if (result > 0 && result < sizeof(buf)) {
+            vstr_add_str(&vfs->root, buf);
+        } else {
+            mp_raise_OSError(GetLastError());
+        }
+        #else
+        if (root[0] != '\0' && root[0] != '/') {
+            char buf[MICROPY_ALLOC_PATH_MAX + 1];
+            const char *cwd = getcwd(buf, sizeof(buf));
+            if (cwd == NULL) {
+                mp_raise_OSError(errno);
```

Comment:
Are you able to write a test which exercises this line?

Reply:
I think the only options are a path that is too long, or a path that the program doesn't have permission to access.  For the latter, maybe `micropython` can be executed when in a root-only directory like `/root`?

Reply:
A long path is the only case I can think of that works without too many platform-specific hoops to jump through. In the additional commit 6815f70, I got it working locally on macOS and Linux. Not on Windows, so the test is skipped there. I’m not sure what’s going on with the CI failures, these seem totally random. Any ideas, before I spend a lot of effort trying to debug something I can’t reproduce locally?

Reply:
Actually, now that I think of it, now that Windows is out anyway (where this wouldn’t work), I could use the failure mode “[ENOENT] A component of the pathname no longer exists.” instead of the too long path. Tried that in e78dc9a, let’s see if that works better. (Before merging, the last two commits should be squashed – I only did it as an additional commit rather than a force-push because I think otherwise the mysterious CI results of 6815f70, which you may still want to look at, might be lost.)

Reply:
That's great you could write such a test, very good thank you!

The CI failure you saw is because `raise` with no arguments is not supported by the native emitter.  Instead do `raise er`.

Reply:
Thanks, that did indeed fix not only the `emit_native_raise_varargs: Assertion 'n_args == 1' failed.` that sounds clear enough in hindsight, but also the more mysterious `TypeError: unsupported types for __lt__: 'dict', 'int'` when compiled without assertions. I had missed that those failures occurred with `--emit native` and could reproduce them that way.

So with that fix, both tests would work. Let me know which one you prefer. The newer one with the deleted directory is probably faster and less susceptible to unexpected external circumstances than the one that creates directories with extremely long names.

Reply:
Yes, the newer one is much better, thanks!